File: backend/schemas/product.py
from pydantic import BaseModel, Field
from uuid import UUID
from decimal import Decimal


class ProductCreate(BaseModel):
    name: str
    description: str
    price: Decimal = Field(gt=0)
    original_price: Decimal | None = Field(default=None, gt=0)
    image: str
    category: str
    rating: float = Field(ge=0, le=5)
    reviews: int = Field(ge=0)
    stock: int = Field(ge=0)
    specs: list[str]
    model: str


class ProductUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    price: Decimal | None = Field(default=None, gt=0)
    original_price: Decimal | None = Field(default=None, gt=0)
    image: str | None = None
    category: str | None = None
    stock: int | None = Field(default=None, ge=0)
    specs: list[str] | None = None
    model: str | None = None


class ProductResponse(BaseModel):
    id: UUID
    name: str
    description: str
    price: Decimal
    original_price: Decimal | None = None
    image: str
    category: str
    rating: float = Field(ge=0, le=5)
    reviews: int = Field(ge=0)
    stock: int = Field(ge=0)
    specs: list[str]
    model: str

File: frontend/types/product.ts
import { User } from "./auth";

export interface Product extends CreateProductType {
  id: string;
  originalPrice?: number | null;
  original_price?: number | null;
  rating: number;
  reviews: number;
}

export interface CreateProductType {
  name: string;
  description: string;
  price: number;
  original_price?: number | null;
  badge?: string;
  category: string;
  specs: string[];
  stock: number;
  image?: string | null;
}

export interface CartItem extends Product {
  quantity: number;
}

export interface CartItem {
  id: string; // CartItem ID from backend
  product: Product; // Product information
  quantity: number; // How many the user wants
}

export interface CartResponse {
  id: string;
  user_id: string;
  items: CartItem[];
}

export interface Order {
  user_detials: User;
  order_items: OrderItem[];
  created_at: string;
  delivery_at: string;
  status: string;
}

export interface OrderItem {
  id: string;
  product_id: string;
  quantity: number;
  product: Product;
}
